pandas/core/sparse/array.py

```diff
@@ -120,298 +135,443 @@ def _wrap_result(name, data, sparse_index, fill_value, dtype=None):
+    @property
+    def values(self):
         """
-        Gets called after any ufunc or other array operations, necessary
-        to pass on the index.
+        Dense values
         """
-        self.sp_index = getattr(obj, 'sp_index', None)
-        self._fill_value = getattr(obj, 'fill_value', None)
+        return self.to_dense()
 
-    def __reduce__(self):
-        """Necessary for making this object picklable"""
-        object_state = list(np.ndarray.__reduce__(self))
-        subclass_state = self.fill_value, self.sp_index
-        object_state[2] = self.sp_values.__reduce__()[2]
-        object_state[2] = (object_state[2], subclass_state)
-        return tuple(object_state)
+    def isna(self):
+        # Two unfortunate things here:
+        # 1. We can't
+        return pd.isna(np.asarray(self))
 
-    def __setstate__(self, state):
-        """Necessary for making this object picklable"""
-        nd_state, own_state = state
-        np.ndarray.__setstate__(self, nd_state)
+    def fillna(self, value=None, method=None, limit=None):
+        """
+        Fill missing values with `value`.
 
-        fill_value, sp_index = own_state[:2]
-        self.sp_index = sp_index
-        self._fill_value = fill_value
+        Parameters
+        ----------
+        value : scalar, optional
+        method : str, optional
 
-    def __len__(self):
-        try:
-            return self.sp_index.length
-        except:
-            return 0
+            .. warning::
 
-    def __unicode__(self):
-        return '{self}\nFill: {fill}\n{index}'.format(
-            self=printing.pprint_thing(self),
-            fill=printing.pprint_thing(self.fill_value),
-            index=printing.pprint_thing(self.sp_index))
+               Using 'method' will result in high memory use,
+               as all `fill_value` methods will be converted to
+               an in-memory ndarray
+        limit : int, optional
 
-    def disable(self, other):
-        raise NotImplementedError('inplace binary ops not supported')
-    # Inplace operators
-    __iadd__ = disable
-    __isub__ = disable
-    __imul__ = disable
-    __itruediv__ = disable
-    __ifloordiv__ = disable
-    __ipow__ = disable
+        Returns
+        -------
+        SparseArray
+
+        Notes
+        -----
+        The result dtype depends on ``self.fill_value``. The goal is
+        to maintain low-memory use. If ``self.fill_value`` is null, the
+        result dtype will be ``SparseDtype(self.dtype, fill_value=value)``.
+        This will preserve amount of memory used before and after filling.
+
+        When ``self.fill_value`` is not NA, the result dtype will be
+        ``SparseDtype(..., fill_value=self.fill_value)``. Again, this
+        preserves the amount of memory used.
+        """
+        # TODO: discussion on what the return type should be.
+        # I think if self.fill_value is NA, then we want to maintain
+        # the sparsity by setting new.fill_value to `value`.
```

Comment:
@jorisvandenbossche I'll push a clarification in a second, but does this still seem strange? To me, it'd be surprising if `fillna(value)` exploded my memory, because I previously had an NA fill value on an array with many missing values.

The user can't (easily) do this on their own. You would be forced to `.astype(SparseDtype(typ, new_fill_value))`, which would be strange, since you're (ab)using astype to fill missing values.

You would also have to always check whether your `fill_value` is NA or not, before knowing whether `.astype` would work like you want.